src/middlewared/middlewared/plugins/alert.py

```diff
@@ -696,69 +741,29 @@ async def __run_alerts(self):
+            )
+            if not locked:
                 self.logger.trace("Running alert source: %r", alert_source.name)
-
                 try:
-                    alerts_a = await self.__run_source(alert_source.name)
+                    this_node_alerts = await self.__run_source(alert_source.name)
                 except UnavailableException:
                     pass
-            for alert in alerts_a:
-                alert.node = master_node
-
-            alerts_b = []
-            if run_on_backup_node and alert_source.run_on_backup_node:
-                try:
-                    alerts_b = [alert
-                                for alert in self.alerts
-                                if alert.node == backup_node and alert.source == alert_source.name]
-                    try:
-                        if not locked:
-                            alerts_b = await self.middleware.call("failover.call_remote", "alert.run_source",
-                                                                  [alert_source.name])
-
-                            alerts_b = [Alert(**dict({k: v for k, v in alert.items()
-                                                      if k in ["args", "datetime", "last_occurrence", "dismissed",
-                                                               "mail"]},
-                                                     klass=AlertClass.class_by_name[alert["klass"]],
-                                                     _source=alert["source"],
-                                                     _key=alert["key"]))
-                                        for alert in alerts_b]
-                    except CallError as e:
-                        if e.errno in [errno.ECONNABORTED, errno.ECONNREFUSED, errno.ECONNRESET, errno.EHOSTDOWN,
-                                       errno.ETIMEDOUT, CallError.EALERTCHECKERUNAVAILABLE]:
-                            pass
-                        else:
-                            raise
-                except ReserveFDException:
-                    self.logger.debug('Failed to reserve a privileged port')
-                except Exception as e:
-                    alerts_b = [
-                        Alert(AlertSourceRunFailedOnBackupNodeAlertClass,
-                              args={
-                                  "source_name": alert_source.name,
-                                  "traceback": str(e),
-                              },
-                              _source=alert_source.name)
-                    ]
 
-            for alert in alerts_b:
-                alert.node = backup_node
+                if run_on_backup_node and alert_source.run_on_backup_node:
+                    other_node_alerts = await self.__run_other_node_alert_source(alert_source.name)
 
-            for alert in alerts_a + alerts_b:
-                self.__handle_alert(alert)
+            for talert, oalert in zip_longest(this_node_alerts, other_node_alerts, fillvalue=None):
```

Comment:
Why do we need to zip them?

Reply:
2 reasons:
1. i'm lazy
2. ergonomics (1 for loop to assign the `alert.node` attribute, and call `__handle_alert(`)

Reply:
Sorry to unresolve this, but I really had a difficulty understanding this code. Seeing `zip` makes me think we do something with pairs of items from `this_node_alerts` and `other_node_alerts`. And it is very confusing, because these alerts are not related.

If you want to do everything in one loop, I would suggest:
```
for alert, node in (
        [(this_node, alert) for alert in this_node_alerts] +
        [(other_node, alert) for alert in other_node_alerts]
):
    alert.node = node
    self.__handle_alert(alert)
```

Reply:
I appreciate the review but I'm not doing as you suggested. I'm trying to remove unnecessary list comprehensions since this is done inside a for loop already and you're adding 2 more. The zip_longest function is perfectly fine here. It's aggregating 2 different iterables (of potentially uneven length).